config/loader.go

```diff
@@ -0,0 +1,202 @@
+package config
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/viper"
+	"golang.org/x/text/cases"
+)
+
+var ErrNilConfig = errors.New("nil configuration")
+
+// IValidatable defines an interface of objects that can self-validate.
+type IValidatable interface {
+	Validate() error
+}
+
+// Loader is responsible for loading config from a source into a struct instance.
+type Loader struct {
+	config    *mapstructure.DecoderConfig
+	exclusive bool
+}
+
+// NewLoader creates a new loader. dest must be a pointer to a struct instance
+// to be populated, other wise, nil returned.
+func NewLoader(dest interface{}) *Loader {
+	loader := &Loader{exclusive: true}
+
+	if err := loader.Init(dest); err != nil {
+		panic(err)
+	}
+
+	return loader
+}
+
+// NewNonExclusiveLoader is just like NewLoader, but the loader returned allows
+// ther to be unknown settings in the source.
+func NewNonExclusiveLoader(dest interface{}) *Loader {
+	loader := &Loader{exclusive: false}
+
+	if err := loader.Init(dest); err != nil {
+		panic(err)
+	}
+	return loader
+}
```

Comment:
> Hm. Good point. I'm familiar with the concept (some of the logging libs use functional options), but it didn't occur to me when writing this. Since there is only one "option", that might actually make things more complicated (certainly in terms of implementation). Something to keep in mind if we need to make the loader more configurable.

👍 

> In our current config, this turned out to be not necessary as the situations this was originally intended for -- configs where values depend on the `backend` specified -- is now handled differently (using `mapstructure` feature to accumulate all "unknown" keys into one filed, and then validating that they do indeed match known backend names). I kept this in case there is a situation where a config is truly free-form and not all valid entries can be predicted. Hopefully, such situations should be exceedingly rare.

SGTM; let's keep it in and see if we need it.  We may want to revisit this in a while and deprecate if that is not the case.